frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -879,22 +883,37 @@ pub mod pallet {
 				Error::<T>::SignedTooMuchWeight,
 			);
 
-			// ensure solution claims is better.
+			// create the submission
+			let reward = T::SignedRewardBase::get();
+			let deposit = Self::deposit_for(&solution, size);
+			let submission = SignedSubmission { who: who.clone(), deposit, reward, solution };
+
+			// insert the submission if the queue has space or it's better than the weakest
+			// eject the weakest if the queue was full
 			let mut signed_submissions = Self::signed_submissions();
-			let ejected_a_solution = signed_submissions.len()
-				== T::SignedMaxSubmissions::get().saturated_into::<usize>();
-			let index = Self::insert_submission(&who, &mut signed_submissions, solution, size)
-				.ok_or(Error::<T>::SignedQueueFull)?;
-
-			// collect deposit. Thereafter, the function cannot fail.
-			let deposit = signed_submissions
-				.get(index)
-				.map(|s| s.deposit)
-				.ok_or(Error::<T>::InvalidSubmissionIndex)?;
-			T::Currency::reserve(&who, deposit).map_err(|_| Error::<T>::SignedCannotPayDeposit)?;
-
-			// store the new signed submission.
-			<SignedSubmissions<T>>::put(signed_submissions);
+			let (inserted, maybe_weakest) = signed_submissions.insert(submission);
+			let ejected_a_solution = maybe_weakest.is_some();
+
+			// it's an error if we neither inserted nor removed any submissions: this indicates
```

Comment:
WDYT? for me this is much easier to reason about, even if it is a few more lines of code: 

```
match (inserted, maybe_weakest)) => {
  (false, None) => { // Not inserted, none removed }
  (false, Some(weakest) => { unreachable!()??? }
  (true, None) => { // inserted and none was removed } 
  (true, Some(weakest) => { // inserted and someone was ejected }
}
```

which can be made into 

match (inserted, maybe_weakest)) => {
  (false, None) => { // Not inserted, none removed }
  (false, Some(weakest) => { unreachable!()??? }
  (true, maybe_removed) => { // inserted, maybe unreserve if maybe_removed.is_some() }
}
```

Reply:
Doesn't that break down into what precisely exists now? We handle the `(false, None)` case with  an `ensure`, don't need to handle a `(false, Some(_))` unreachable case, and then handle the `(true, maybe_removed)` case inline. Agree that those are the cases we need to consider; I just think that the existing formulation is easier to read.

Reply:
I read this in the downstream PR again and it was still confusing to me. Probably our definitions of "easy to read" differs. For me a readable code is one that can be easily understood with the least amount of time spent reading it. This entails being slightly verbose sometimes. 

A match statement's basis is that it cannot have any unhandled cases, and the reader can clearly see each case. The sequence of `if` statements took more time to reason about, at least in my case. 

I leave it up to you though, just wanted to share my detailed opinion.